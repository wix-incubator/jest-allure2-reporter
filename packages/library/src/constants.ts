export const PREFIX = 'allure2' as const;
export const ALLURE_ID = [PREFIX, 'allureId'] as const;
export const DESCRIPTION = [PREFIX, 'description'] as const;
export const DESCRIPTION_HTML = [PREFIX, 'descriptionHtml'] as const;
export const ISSUE = [PREFIX, 'issue'] as const;
export const LEAD = [PREFIX, 'lead'] as const;
export const LINK = [PREFIX, 'link'] as const;
export const OWNER = [PREFIX, 'owner'] as const;
export const SEVERITY = [PREFIX, 'severity'] as const;
export const TMS_LINK = [PREFIX, 'tmsLink'] as const;
export const TAG = [PREFIX, 'tags'] as const;
