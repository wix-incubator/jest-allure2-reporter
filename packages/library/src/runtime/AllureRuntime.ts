/* eslint-disable @typescript-eslint/ban-types, prefer-rest-params */
import type { Metadata } from 'jest-metadata';
import { Status, Stage } from '@noomorph/allure-js-commons';
import type {
  StatusDetails,
  Parameter,
  ParameterOptions,
  LabelName,
} from '@noomorph/allure-js-commons';

import {
  CURRENT_STEP,
  DESCRIPTION,
  DESCRIPTION_HTML,
  LABELS,
  LINKS,
  PREFIX,
} from '../constants';
import { isPromiseLike } from '../utils/isPromiseLike';
import type { AllureTestStepMetadata } from '../metadata/metadata';

export type AllureRuntimeConfig = {
  metadataProvider: () => Metadata;
  nowProvider: () => number;
};

export type ParameterOrString = string | Parameter;

export class AllureRuntime {
  readonly #metadataProvider: AllureRuntimeConfig['metadataProvider'];
  readonly #now: AllureRuntimeConfig['nowProvider'];

  get #metadata(): Metadata {
    return this.#metadataProvider();
  }

  constructor(config: AllureRuntimeConfig) {
    this.#metadataProvider = config.metadataProvider;
    this.#now = config.nowProvider;
  }

  description(value: string) {
    this.#metadata.push(DESCRIPTION, [value]);
  }

  descriptionHtml(value: string) {
    this.#metadata.push(DESCRIPTION_HTML, [value]);
  }

  label(name: LabelName | string, value: string) {
    this.#metadata.push(LABELS, [{ name, value }]);
  }

  link(name: string, url: string, type?: string) {
    this.#metadata.push(LINKS, [{ name, url, type }]);
  }

  parameter(name: string, value: unknown, options?: ParameterOptions) {
    this.#metadata.push(this.#localPath('parameters'), [
      {
        name,
        value: `${value}`,
        ...options,
      },
    ]);
  }

  parameters(parameters: Record<string, unknown>) {
    for (const [name, value] of Object.entries(parameters)) {
      this.parameter(name, value);
    }
  }

  attachment(name: string, content: Buffer | string, type: string) {
    this.#metadata.push(this.#localPath('attachments'), [
      { name, content, type },
    ]);
  }

  createStep<F extends Function>(
    name: string,
    arguments_: ParameterOrString[],
    function_: F,
  ): F;
  createStep<F extends Function>(name: string, function_: F): F;
  createStep<F extends Function>(
    name: string,
    maybeArguments: F | ParameterOrString[],
    maybeFunction?: F,
  ): F {
    // eslint-disable-next-line @typescript-eslint/no-this-alias,unicorn/no-this-assignment
    const runtime = this;
    const function_ =
      typeof maybeArguments === 'function' ? maybeArguments : maybeFunction!;
    const definitions = Array.isArray(maybeArguments) ? maybeArguments : null;

    function wrapped(this: unknown) {
      const thisWrapped = () => Reflect.apply(function_, this, arguments);
      return runtime.step(name, () => {
        const parameters =
          definitions ?? Array.from(arguments, (_, index) => `${index}`);

        // eslint-disable-next-line unicorn/no-for-loop
        for (let index = 0; index < parameters.length; index++) {
          const parameter = parameters[index];
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { name, value: _value, ...options } = parameter as Parameter;
          runtime.parameter(name ?? parameter, arguments[index], options);
        }
        return thisWrapped();
      });
    }

    wrapped.toString = function_.toString.bind(function_);
    return wrapped as unknown as F;
  }

  step(name: string, function_: () => Promise<unknown>): Promise<unknown>;
  step(name: string, function_: () => unknown): unknown;
  step(name: string, function_: () => unknown) {
    this.#startStep(name, function_);
    const end = this.#stopStep;

    let result: unknown;
    try {
      result = function_();

      if (isPromiseLike(result)) {
        this.#updateStep(Stage.RUNNING);

        result.then(
          () => end(Status.PASSED),
          (error) =>
            end(Status.FAILED, { message: error.message, trace: error.stack }),
        );
      } else {
        end(Status.PASSED);
      }

      return result;
    } catch (error: unknown) {
      end(Status.FAILED, {
        message: (error as Error).message,
        trace: (error as Error).stack,
      });
      throw error;
    }
  }

  // eslint-disable-next-line @typescript-eslint/ban-types
  #startStep = (name: string, function_: Function) => {
    const count = this.#metadata.get(this.#localPath('steps', 'length'), 0);
    this.#metadata.push(this.#localPath('steps'), [
      {
        name,
        stage: Stage.SCHEDULED,
        start: this.#now(),
        code: function_.toString(),
      },
    ]);
    // eslint-disable-next-line unicorn/no-array-push-push
    this.#metadata.push(CURRENT_STEP, ['steps', `${count}`]);
  };

  #stopStep = (status: Status, statusDetails?: StatusDetails) => {
    this.#metadata.assign(this.#localPath(), {
      stage: Stage.FINISHED,
      status,
      statusDetails,
      stop: this.#now(),
    });
    const currentStep = this.#metadata.get(CURRENT_STEP, []) as string[];
    this.#metadata.set(CURRENT_STEP, currentStep.slice(0, -2));
  };

  #updateStep = (stage: Stage) => {
    this.#metadata.set(this.#localPath('stage'), stage);
  };

  #localPath(key?: keyof AllureTestStepMetadata, ...innerKeys: string[]) {
    const stepPath = this.#metadata.get(CURRENT_STEP, []) as string[];
    const allKeys = key ? [key, ...innerKeys] : innerKeys;
    return [PREFIX, ...stepPath, ...allKeys];
  }
}
