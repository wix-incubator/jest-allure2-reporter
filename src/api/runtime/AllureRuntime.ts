import type { Metadata } from 'jest-metadata';
import type { LabelName, Status, StatusDetails } from 'jest-allure2-reporter';

import type { ParameterOrString } from './index';
import { isPromiseLike } from '../../utils/isPromiseLike';
import { wrapFunction } from '../../utils/wrapFunction';
import { formatString } from '../../utils/formatString';

import type {
  AllureRuntimeBindOptions,
  IAllureRuntime,
  ParameterOptions,
} from './IAllureRuntime';
import { AllureTestCaseMetadataProxy } from './proxies';

export type AllureRuntimeConfig = {
  metadataProvider: () => Metadata;
  nowProvider: () => number;
};

const constant =
  <T>(value: T) =>
  () =>
    value;
const noop = (..._arguments: unknown[]) => void 0;

export class AllureRuntime implements IAllureRuntime {
  readonly #metadataProvider: AllureRuntimeConfig['metadataProvider'];
  readonly #now: AllureRuntimeConfig['nowProvider'];
  #idle = Promise.resolve();

  get #metadata(): AllureTestCaseMetadataProxy {
    return new AllureTestCaseMetadataProxy(this.#metadataProvider());
  }

  constructor(config: AllureRuntimeConfig) {
    this.#metadataProvider = config.metadataProvider;
    this.#now = config.nowProvider;
  }

  $bind(options?: AllureRuntimeBindOptions): AllureRuntime {
    const { metadata = true, time = false } = options ?? {};

    return new AllureRuntime({
      metadataProvider: metadata
        ? constant(this.#metadataProvider())
        : this.#metadataProvider,
      nowProvider: time ? constant(this.#now()) : this.#now,
    });
  }

  $plug(_callback: any): any {
    // TODO: implement
  }

  async flush(): Promise<void> {
    await this.#idle;
  }

  description(value: string) {
    this.#metadata.push('description', [value]);
  }

  descriptionHtml(value: string) {
    this.#metadata.push('descriptionHtml', [value]);
  }

  label(name: LabelName | string, value: string) {
    this.#metadata.push('labels', [{ name, value }]);
  }

  link(url: string, name = url, type?: string) {
    this.#metadata.push('links', [{ name, url, type }]);
  }

  parameter(name: string, value: unknown, options?: ParameterOptions) {
    this.#metadata.push('parameters', [
      {
        name,
        value: `${value}`,
        ...options,
      },
    ]);
  }

  parameters(parameters: Record<string, unknown>) {
    for (const [name, value] of Object.entries(parameters)) {
      this.parameter(name, value);
    }
  }

  status(status: Status, statusDetails?: StatusDetails) {
    if (statusDetails) {
      this.#metadata.assign({ status, statusDetails });
    } else {
      this.#metadata.assign({ statusDetails });
    }
  }

  statusDetails(statusDetails: StatusDetails) {
    this.#metadata.set('statusDetails', statusDetails);
  }

  step<T = unknown>(name: string, function_: () => T): T {
    this.#startStep(name, function_);
    const end = this.#stopStep;

    let result: T;
    try {
      result = function_();

      if (isPromiseLike(result)) {
        this.#metadata.set('stage', 'running');

        result.then(
          () => end('passed'),
          (error) =>
            end('failed', { message: error.message, trace: error.stack }),
        );
      } else {
        end('passed');
      }

      return result;
    } catch (error: unknown) {
      end('failed', {
        message: (error as Error).message,
        trace: (error as Error).stack,
      });
      throw error;
    }
  }

  createStep<F extends Function>(
    nameFormat: string,
    maybeParameters: F | ParameterOrString[],
    maybeFunction?: F,
  ): F {
    const function_: any = maybeFunction ?? (maybeParameters as F);
    if (typeof function_ !== 'function') {
      throw new TypeError(
        `Expected a function, got ${typeof function_} instead`,
      );
    }

    const runtime = this;
    const userParameters = Array.isArray(maybeParameters)
      ? maybeParameters
      : null;

    const handleArguments = (arguments_: IArguments) => {
      const parameters = userParameters ?? Array.from(arguments_, noop);
      const allureParameters = parameters.map(resolveParameter, arguments_);
      for (const [name, value, options] of allureParameters) {
        runtime.parameter(name, value, options);
      }
    };

    return wrapFunction(function_, function (this: unknown) {
      const arguments_ = arguments;
      const name = formatString(nameFormat, ...arguments_);
      return runtime.step(
        name,
        wrapFunction(
          function_,
          function step(this: unknown) {
            handleArguments(arguments_);
            return Reflect.apply(function_, this, arguments_);
          }.bind(this),
        ),
      );
    });
  }

  #startStep = (name: string, function_: Function) => {
    this.#metadata.$startStep().assign({
      stage: 'scheduled',
      start: this.#now(),
      code: function_.toString(),
      description: [name],
    });
  };

  #stopStep = (status: Status, statusDetails?: StatusDetails) => {
    const existing = this.#metadata.get(undefined, {} as any);
    this.#metadata
      .assign({
        stage: 'finished',
        status: existing.status ?? status,
        statusDetails: existing.statusDetails ?? statusDetails,
        stop: this.#now(),
      })
      .$stopStep();
  };
}

function resolveParameter(
  this: unknown[],
  parameter: ParameterOrString | undefined,
  index: number,
) {
  const { name = `${index}`, ...options } =
    typeof parameter === 'string' ? { name: parameter } : parameter ?? {};

  return [name, this[index], options] as const;
}
