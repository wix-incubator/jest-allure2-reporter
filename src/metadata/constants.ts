export const PREFIX = 'allure2' as const;

export const SHARED_CONFIG = [PREFIX, 'config'] as const;

export const CODE = [PREFIX, 'code'] as const;
export const WORKER_ID = [PREFIX, 'workerId'] as const;
export const HIDDEN = [PREFIX, 'hidden'] as const;

export const START = [PREFIX, 'start'] as const;

export const STOP = [PREFIX, 'stop'] as const;

export const CURRENT_STEP = [PREFIX, 'currentStep'] as const;
export const DESCRIPTION = [PREFIX, 'description'] as const;
export const DESCRIPTION_HTML = [PREFIX, 'descriptionHtml'] as const;
export const LABELS = [PREFIX, 'labels'] as const;
export const LINKS = [PREFIX, 'links'] as const;
