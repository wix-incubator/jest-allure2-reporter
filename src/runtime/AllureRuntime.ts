import path from 'node:path';

import type {
  LabelName,
  ParameterOptions,
  StatusDetails,
} from '@noomorph/allure-js-commons';
import { Stage, Status } from '@noomorph/allure-js-commons';
import type { Metadata } from 'jest-metadata';
import type {
  AllureTestStepMetadata,
  AttachmentContent,
  AttachmentOptions,
  IAllureRuntime,
  ParameterOrString,
} from 'jest-allure2-reporter';

import {
  CURRENT_STEP,
  DESCRIPTION,
  DESCRIPTION_HTML,
  LABELS,
  LINKS,
  PREFIX,
} from '../constants';
import { isPromiseLike } from '../utils/isPromiseLike';
import { inferMimeType } from '../utils/inferMimeType';
import { hijackFunction } from '../utils/hijackFunction';
import type { Function_, MaybePromise } from '../utils/types';
import { processMaybePromise } from '../utils/processMaybePromise';
import { wrapFunction } from '../utils/wrapFunction';
import { formatString } from '../utils/formatString';

import type { IAttachmentsHandler } from './AttachmentsHandler';

export type AllureRuntimeConfig = {
  attachmentsHandler: IAttachmentsHandler;
  metadataProvider: () => Metadata;
  nowProvider: () => number;
};

const noop = (..._arguments: unknown[]) => void 0;

export class AllureRuntime implements IAllureRuntime {
  readonly #attachmentsHandler: AllureRuntimeConfig['attachmentsHandler'];
  readonly #metadataProvider: AllureRuntimeConfig['metadataProvider'];
  readonly #now: AllureRuntimeConfig['nowProvider'];
  #idle = Promise.resolve();

  get #metadata(): Metadata {
    return this.#metadataProvider();
  }

  constructor(config: AllureRuntimeConfig) {
    this.#attachmentsHandler = config.attachmentsHandler;
    this.#metadataProvider = config.metadataProvider;
    this.#now = config.nowProvider;
  }

  async flush(): Promise<void> {
    await this.#idle;
  }

  description(value: string) {
    this.#metadata.push(DESCRIPTION, [value]);
  }

  descriptionHtml(value: string) {
    this.#metadata.push(DESCRIPTION_HTML, [value]);
  }

  label(name: LabelName | string, value: string) {
    this.#metadata.push(LABELS, [{ name, value }]);
  }

  link(name: string, url: string, type?: string) {
    this.#metadata.push(LINKS, [{ name, url, type }]);
  }

  parameter(name: string, value: unknown, options?: ParameterOptions) {
    this.#metadata.push(this.#localPath('parameters'), [
      {
        name,
        value: `${value}`,
        ...options,
      },
    ]);
  }

  parameters(parameters: Record<string, unknown>) {
    for (const [name, value] of Object.entries(parameters)) {
      this.parameter(name, value);
    }
  }

  status(status?: Status | StatusDetails, statusDetails?: StatusDetails) {
    this.#metadata.assign(this.#localPath(), { status, statusDetails });
  }

  statusDetails(statusDetails: StatusDetails) {
    this.#metadata.assign(this.#localPath(), { statusDetails });
  }

  attachment<T extends AttachmentContent>(
    name: string,
    content: MaybePromise<T>,
    mimeType?: string,
  ): typeof content {
    return processMaybePromise(
      content,
      this.#handleDynamicAttachment({ name, mimeType }),
    );
  }

  createAttachment<T extends AttachmentContent>(
    function_: Function_<MaybePromise<T>>,
    rawOptions: string | AttachmentOptions,
  ): typeof function_ {
    const options = this.#resolveAttachmentOptions(rawOptions);
    return hijackFunction(function_, this.#handleDynamicAttachment(options));
  }

  fileAttachment(
    filePathOrPromise: MaybePromise<string>,
    rawOptions?: string | AttachmentOptions,
  ): any {
    const options = this.#resolveAttachmentOptions(rawOptions);
    return processMaybePromise(
      filePathOrPromise,
      this.#handleStaticAttachment(options),
    );
  }

  createFileAttachment(
    function_: Function_<MaybePromise<string>>,
    rawOptions?: string | AttachmentOptions,
  ): typeof function_ {
    const options = this.#resolveAttachmentOptions(rawOptions);
    return hijackFunction<string>(
      function_,
      this.#handleStaticAttachment(options),
    );
  }

  step<T = unknown>(name: string, function_: () => T): T {
    this.#startStep(name, function_);
    const end = this.#stopStep;

    let result: T;
    try {
      result = function_();

      if (isPromiseLike(result)) {
        this.#updateStep(Stage.RUNNING);

        result.then(
          () => end(Status.PASSED),
          (error) =>
            end(Status.FAILED, { message: error.message, trace: error.stack }),
        );
      } else {
        end(Status.PASSED);
      }

      return result;
    } catch (error: unknown) {
      end(Status.FAILED, {
        message: (error as Error).message,
        trace: (error as Error).stack,
      });
      throw error;
    }
  }

  createStep<F extends Function>(
    nameFormat: string,
    maybeParameters: F | ParameterOrString[],
    maybeFunction?: F,
  ): F {
    const function_: any = maybeFunction ?? (maybeParameters as F);
    if (typeof function_ !== 'function') {
      throw new TypeError(
        `Expected a function, got ${typeof function_} instead`,
      );
    }

    const runtime = this;
    const userParameters = Array.isArray(maybeParameters)
      ? maybeParameters
      : null;

    const handleArguments = (arguments_: IArguments) => {
      const parameters = userParameters ?? Array.from(arguments_, noop);
      const allureParameters = parameters.map(resolveParameter, arguments_);
      for (const [name, value, options] of allureParameters) {
        runtime.parameter(name, value, options);
      }
    };

    return wrapFunction(function_, function (this: unknown) {
      const arguments_ = arguments;
      const name = formatString(nameFormat, ...arguments_);
      return runtime.step(
        name,
        wrapFunction(
          function_,
          function step(this: unknown) {
            handleArguments(arguments_);
            return Reflect.apply(function_, this, arguments_);
          }.bind(this),
        ),
      );
    });
  }

  #startStep = (name: string, function_: Function) => {
    const count = this.#metadata.get(this.#localPath('steps', 'length'), 0);
    this.#metadata.push(this.#localPath('steps'), [
      {
        name,
        stage: Stage.SCHEDULED,
        start: this.#now(),
        code: function_.toString(),
      },
    ]);
    // eslint-disable-next-line unicorn/no-array-push-push
    this.#metadata.push(CURRENT_STEP, ['steps', `${count}`]);
  };

  #stopStep = (status: Status, statusDetails?: StatusDetails) => {
    const existing = this.#metadata.get(this.#localPath(), {} as any);

    this.#metadata.assign(this.#localPath(), {
      stage: Stage.FINISHED,
      status: existing.status ?? status,
      statusDetails: existing.statusDetails ?? statusDetails,
      stop: this.#now(),
    });

    const currentStep = this.#metadata.get(CURRENT_STEP, []) as string[];
    this.#metadata.set(CURRENT_STEP, currentStep.slice(0, -2));
  };

  #updateStep = (stage: Stage) => {
    this.#metadata.set(this.#localPath('stage'), stage);
  };

  #localPath(key?: keyof AllureTestStepMetadata, ...innerKeys: string[]) {
    const stepPath = this.#metadata.get(CURRENT_STEP, []) as string[];
    const allKeys = key ? [key, ...innerKeys] : innerKeys;
    return [PREFIX, ...stepPath, ...allKeys];
  }

  #resolveAttachmentOptions(rawOptions?: string | AttachmentOptions) {
    return !rawOptions || typeof rawOptions === 'string'
      ? { name: rawOptions, mimeType: undefined }
      : rawOptions;
  }

  #handleDynamicAttachment =
    ({ name: nameFormat, mimeType }: AttachmentOptions) =>
    (content: AttachmentContent, arguments_?: unknown[]) => {
      const name = this.#formatName(nameFormat, arguments_);
      const source = this.#attachmentsHandler.placeAttachment(name, content);
      this.#metadata.push(this.#localPath('attachments'), [
        {
          name,
          source: path.resolve(source),
          type: mimeType ?? inferMimeType(name),
        },
      ]);
      const promise = this.#attachmentsHandler.writeAttachment(source, content);
      this.#idle = this.#idle.then(() => promise);
    };

  #handleStaticAttachment =
    ({ name: rawName, mimeType }: AttachmentOptions) =>
    (filePath: string, arguments_?: unknown[]) => {
      const name = this.#formatName(
        rawName ?? path.basename(filePath),
        arguments_,
      );
      const securedPath = this.#attachmentsHandler.placeAttachment(name);
      const result = this.#attachmentsHandler.secureAttachment(
        filePath,
        securedPath,
      );
      this.#metadata.push(this.#localPath('attachments'), [
        {
          name,
          source: path.resolve(result.destinationPath),
          type: mimeType ?? inferMimeType(name, filePath),
        },
      ]);
      if (result.promise) {
        this.#idle = this.#idle.then(() => result.promise);
      }
    };

  #formatName(nameFormat?: string, arguments_?: unknown[]) {
    return arguments_
      ? formatString(nameFormat ?? 'untitled', ...arguments_)
      : nameFormat ?? 'untitled';
  }
}

function resolveParameter(
  this: unknown[],
  parameter: ParameterOrString | undefined,
  index: number,
) {
  const { name = `${index}`, ...options } =
    typeof parameter === 'string' ? { name: parameter } : parameter ?? {};

  return [name, this[index], options] as const;
}
