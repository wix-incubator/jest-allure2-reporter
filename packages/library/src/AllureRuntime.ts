import type {Metadata} from 'jest-metadata';
import {$POINTER, PREFIX} from "./constants";
import type {AllureTestStepMetadata} from "./options";
import {Status, Stage} from "@noomorph/allure-js-commons";
import type {StatusDetails} from "@noomorph/allure-js-commons";
import {isPromiseLike} from "./utils/isPromiseLike";

export class AllureRuntime {
  readonly #metadataProvider: () => Metadata;

  get #metadata(): Metadata {
    return this.#metadataProvider();
  }

  constructor(metadataProvider: () => Metadata) {
    this.#metadataProvider = metadataProvider;
  }

  attachment(name: string, content: Buffer | string, type: string) {
    this.#metadata.push(this.#localPath('attachments'), [{ name, content, type }]);
  }

  step(name: string, fn: () => Promise<unknown>): Promise<unknown>;
  step(name: string, fn: () => unknown): unknown;
  step(name: string, fn: () => unknown) {
    this.#startStep(name);
    const end = this.#stopStep;

    let result: unknown;
    try {
      result = fn();

      if (isPromiseLike(result)) {
        this.#updateStep(Stage.RUNNING);

        result.then(
          () => end(Status.PASSED),
          (error) => end(Status.FAILED, { message: error.message, trace: error.stack }),
        );
      } else {
        end(Status.PASSED);
      }

      return result;
    } catch (error: unknown) {
      end(Status.FAILED, { message: (error as Error).message, trace: (error as Error).stack });
      throw error;
    }
  }

  #startStep = (name: string) => {
    const count = this.#metadata.get(this.#localPath('steps', 'length'), 0);
    this.#metadata.push(this.#localPath('steps'), [{ name, stage: Stage.SCHEDULED }]);
    this.#metadata.push($POINTER, ['steps', `${count}`]);
  };

  #stopStep = (status: Status, statusDetails?: StatusDetails) => {
    this.#metadata.assign(this.#localPath(), {
      stage: Stage.FINISHED,
      status,
      statusDetails,
    });
    const $pointer = this.#metadata.get($POINTER, []) as string[];
    this.#metadata.set($POINTER, $pointer.slice(0, -2));
  };

  #updateStep = (stage: Stage) => {
    this.#metadata.set(this.#localPath('stage'), stage);
  };

  #localPath(key?: keyof AllureTestStepMetadata, ...innerKeys: string[]) {
    const stepPath = this.#metadata.get($POINTER, []) as string[];
    const allKeys = key ? [key, ...innerKeys] : innerKeys;
    return [PREFIX, ...stepPath, ...allKeys];
  }
}
