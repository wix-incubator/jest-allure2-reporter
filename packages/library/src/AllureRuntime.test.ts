import {AllureRuntime} from "./AllureRuntime";
import {state} from 'jest-metadata';
import {PREFIX} from "./constants";

describe('AllureRuntime', () => {
  it('should add attachments within the steps', async () => {
    const runtime = new AllureRuntime(() => state);
    runtime.attachment('attachment1', Buffer.from('first'), 'text/plain');
    await runtime.step('outer step', async () => {
      try {
        runtime.step('inner step 1', () => {
          runtime.attachment('attachment2', 'second', 'text/plain');

          const err = new Error('Sync error');
          err.stack = 'Test stack';
          throw err;
        });
      } catch {}
      runtime.step('inner step 2', () => {});
      runtime.attachment('attachment3', 'third', 'text/plain');
      await runtime.step('inner step 3', async () => {
        runtime.attachment('attachment4', 'fourth', 'text/plain');

        const err = new Error('Async error');
        err.stack = 'Test stack';
        throw err;
      }).catch(() => {});
    });
    runtime.attachment('attachment5', Buffer.from('fifth'), 'text/plain');
    expect(state.get(PREFIX)).toMatchSnapshot();
  });
});
