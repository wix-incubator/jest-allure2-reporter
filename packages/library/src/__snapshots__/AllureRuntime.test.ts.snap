// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllureRuntime should add attachments within the steps 1`] = `
{
  "$pointer": [],
  "attachments": [
    {
      "content": {
        "data": [
          102,
          105,
          114,
          115,
          116,
        ],
        "type": "Buffer",
      },
      "name": "attachment1",
      "type": "text/plain",
    },
    {
      "content": {
        "data": [
          102,
          105,
          102,
          116,
          104,
        ],
        "type": "Buffer",
      },
      "name": "attachment5",
      "type": "text/plain",
    },
  ],
  "steps": [
    {
      "attachments": [
        {
          "content": "third",
          "name": "attachment3",
          "type": "text/plain",
        },
      ],
      "name": "outer step",
      "stage": "finished",
      "status": "passed",
      "statusDetails": undefined,
      "steps": [
        {
          "attachments": [
            {
              "content": "second",
              "name": "attachment2",
              "type": "text/plain",
            },
          ],
          "name": "inner step 1",
          "stage": "finished",
          "status": "failed",
          "statusDetails": {
            "message": "Sync error",
            "trace": "Test stack",
          },
        },
        {
          "name": "inner step 2",
          "stage": "finished",
          "status": "passed",
          "statusDetails": undefined,
        },
        {
          "attachments": [
            {
              "content": "fourth",
              "name": "attachment4",
              "type": "text/plain",
            },
          ],
          "name": "inner step 3",
          "stage": "finished",
          "status": "failed",
          "statusDetails": {
            "message": "Async error",
            "trace": "Test stack",
          },
        },
      ],
    },
  ],
}
`;
